### DISCUSSION TITLE
DropwizardResourceConfig Logs Paths that Do Not Work

### DISCUSSION THREAD

**Discussion by JKomoroski** [dropwizard/dropwizard#9345] [General]
During application startup applications get a log with a list of paths for configured resources:
```
INFO  [2024-06-26 23:08:21,960] io.dropwizard.jersey.DropwizardResourceConfig: The following paths were found for the configured resources:

    GET     /nested/path (ski.komoro.example.resources.NestedResource)
    GET     /nested/path/ping (ski.komoro.example.resources.PingResource)
    GET     /nested/path/ping2 (ski.komoro.example.resources.Ping2Resource)
```

However, if those resources conflict or jersey does not handle them, the resources return a 404 at run time.
Example Access Logs:
```
127.0.0.1 - - [26/Jun/2024:23:08:22 +0000] "GET /nested/path HTTP/1.1" 404 43 "-" "Jersey/3.0.12 (Apache HttpClient 5.3.1)" 38
127.0.0.1 - - [26/Jun/2024:23:08:22 +0000] "GET /nested/path/ping HTTP/1.1" 200 4 "-" "Jersey/3.0.12 (Apache HttpClient 5.3.1)" 8
127.0.0.1 - - [26/Jun/2024:23:08:22 +0000] "GET /nested/path/ping2 HTTP/1.1" 200 4 "-" "Jersey/3.0.12 (Apache HttpClient 5.3.1)" 2
```
In larger applications, it can be hard to tell exactly which resources are at fault for the error and seeing the misconfigured endpoints in the startup logs makes things difficult.

I would prefer if the application threw an error on startup if resources are not configured properly, but given that is likely the domain of Jersey, it would be nice if the DropwizardResourceConfig class did not log resources that do not work. I do not know if that is possible, but perhaps jersey has an api that exports all registered endpoints?

I've setup a sample application based on the DW Archetype, which will reproduce the stated issue during a `mvn clean install` and fail an integration test which attempts to do a GET on the logged resources. 
https://github.com/JKomoroski/resource-log-example

Reviewing the 3 resource classes in the project is an easy way to see how I've intentionally misconfigured the application resource classes.

**Comment by zUniQueX** [comment]
Hi @JKomoroski. The 'Jersey-way' would be to use [sub-resources](https://eclipse-ee4j.github.io/jersey.github.io/documentation/latest/jaxrs-resources.html) in your application to avoid conflicting paths at different root-resources. The logged endpoints by dropwizard are _declared_ endpoints, not _handled_ endpoints. If Jersey doesn't want to handle an endpoint, it was still declared by the application. If the implementation changes to handle those endpoints, we shouldn't filter them in our code.

If I understand you correctly, a non-working endpoint is a Jersey `Resource` conflicting with the path of an other `Resource`. You could implement a detector for this behavior somewhat like this:
```java
public class OverlappingPathValidator extends AbstractResourceModelVisitor {

    private static final Logger LOGGER = LoggerFactory.getLogger(OverlappingPathValidator.class);

    private final Map<String, Resource> resourceMap = new HashMap<>();

    @Override
    public void visitResource(Resource resource) {
        validateResource(resource);
        resource.getChildResources().forEach(childResource -> childResource.accept(this));
    }

    @Override
    public void visitChildResource(Resource resource) {
        validateResource(resource);
        resource.getChildResources().forEach(childResource -> childResource.accept(this));
    }

    private void validateResource(Resource resource) {
        Resource parent = resource;
        StringBuilder fullPath = new StringBuilder(resource.getPath());
        while (parent.getParent() != null) {
            parent = parent.getParent();
            fullPath.insert(0, parent.getPath());
        }
        Resource previous = resourceMap.put(fullPath.toString(), resource);
        if (previous != null && previous != resource) {
            LOGGER.error("Path {} is mapped by multiple resources", fullPath);
        }
    }

    public static class OverlappingPathValidatorRegistrar implements ApplicationEventListener {

        @Override
        public void onEvent(ApplicationEvent event) {
            if (event.getType() == ApplicationEvent.Type.INITIALIZATION_APP_FINISHED) {
                OverlappingPathValidator overlappingPathValidator = new OverlappingPathValidator();
                event.getResourceModel().getResources().forEach(resource -> resource.accept(overlappingPathValidator));
            }
        }

        @Override
        public RequestEventListener onRequest(RequestEvent requestEvent) {
            return null;
        }
    }
}
```

When registering this listener with `environment.jersey().register(OverlappingPathValidator.OverlappingPathValidatorRegistrar.class)` in your application, it logs the following line:
```
ERROR [2024-09-07 15:23:20,504] ski.komoro.example.util.OverlappingPathValidator: Path /nested/path is mapped by multiple resources
```

This should help rewriting the code to avoid overlapping paths in root-resources.

**Comment by JKomoroski** [comment]
Thanks for the thoughtful and detailed response. I'll play with this code and see how it works for me. I suspect throwing an exception at startup would be a better experience for my use.

My trouble is that I'm working in a large monolithic application with deeply nested paths, a hundred resource classes and a multiple of that in actual endpoints. We use subresources and sub-resource locators already but not particularly well or constantly. The process of having a new developer add new endpoint often results in a misplaced resource similar to the example I provided. When the dev sends me the start up log, I often have to point out that I the startup log is not reliable, negating some of the value it offers.
